### ISSUE TITLE
'PEM_read_bio_RSAPrivateKey' is deprecated: Since OpenSSL 3.0

### ISSUE THREAD

**Issue by dhardyuk** [obuno/gcmlxc#1]
Hi,

I've worked through your excellent guide on setting Guacamole up in an Ubuntu based container on Proxmox. I've used the 22.10 Ubuntu version and have hit some snags.

Before I go again using the 22.04 Ubuntu release is there an easy fix to these errors?

`Making all in .
make[3]: Entering directory '/usr/src/guacamole-server-1.4.0/src/common-ssh'
  CC       libguac_common_ssh_la-key.lo
key.c: In function 'guac_common_ssh_key_alloc':
key.c:63:9: error: 'PEM_read_bio_RSAPrivateKey' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
   63 |         rsa_key = PEM_read_bio_RSAPrivateKey(key_bio, NULL, NULL, passphrase);
      |         ^~~~~~~
In file included from key.c:33:
/usr/include/openssl/pem.h:447:1: note: declared here
  447 | DECLARE_PEM_rw_cb_attr(OSSL_DEPRECATEDIN_3_0, RSAPrivateKey, RSA)
      | ^~~~~~~~~~~~~~~~~~~~~~
key.c:79:9: error: 'RSA_get0_key' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
   79 |         RSA_get0_key(rsa_key, &key_n, &key_e, NULL);
      |         ^~~~~~~~~~~~
In file included from common-ssh/rsa-compat.h:26,
                 from key.c:25:
/usr/include/openssl/rsa.h:217:28: note: declared here
  217 | OSSL_DEPRECATEDIN_3_0 void RSA_get0_key(const RSA *r,
      |                            ^~~~~~~~~~~~
key.c:105:9: error: 'PEM_read_bio_DSAPrivateKey' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  105 |         dsa_key = PEM_read_bio_DSAPrivateKey(key_bio, NULL, NULL, passphrase);
      |         ^~~~~~~
/usr/include/openssl/pem.h:453:1: note: declared here
  453 | DECLARE_PEM_rw_cb_attr(OSSL_DEPRECATEDIN_3_0, DSAPrivateKey, DSA)
      | ^~~~~~~~~~~~~~~~~~~~~~
key.c:121:9: error: 'DSA_get0_pqg' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  121 |         DSA_get0_pqg(dsa_key, &key_p, &key_q, &key_g);
      |         ^~~~~~~~~~~~
In file included from common-ssh/dsa-compat.h:26,
                 from key.c:23:
/usr/include/openssl/dsa.h:201:28: note: declared here
  201 | OSSL_DEPRECATEDIN_3_0 void DSA_get0_pqg(const DSA *d, const BIGNUM **p,
      |                            ^~~~~~~~~~~~
key.c:122:9: error: 'DSA_get0_key' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  122 |         DSA_get0_key(dsa_key, &pub_key, NULL);
      |         ^~~~~~~~~~~~
/usr/include/openssl/dsa.h:204:28: note: declared here
  204 | OSSL_DEPRECATEDIN_3_0 void DSA_get0_key(const DSA *d, const BIGNUM **pub_key,
      |                            ^~~~~~~~~~~~
key.c: In function 'guac_common_ssh_key_free':
key.c:164:9: error: 'RSA_free' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  164 |         RSA_free(key->rsa);
      |         ^~~~~~~~
/usr/include/openssl/rsa.h:293:28: note: declared here
  293 | OSSL_DEPRECATEDIN_3_0 void RSA_free(RSA *r);
      |                            ^~~~~~~~
key.c:166:9: error: 'DSA_free' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  166 |         DSA_free(key->dsa);
      |         ^~~~~~~~
/usr/include/openssl/dsa.h:127:28: note: declared here
  127 | OSSL_DEPRECATEDIN_3_0 void DSA_free(DSA *r);
      |                            ^~~~~~~~
key.c: In function 'guac_common_ssh_key_sign':
key.c:202:13: error: 'RSA_sign' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  202 |             if (RSA_sign(NID_sha1, digest, dlen, sig, &len, key->rsa) == 1)
      |             ^~
/usr/include/openssl/rsa.h:348:27: note: declared here
  348 | OSSL_DEPRECATEDIN_3_0 int RSA_sign(int type, const unsigned char *m,
      |                           ^~~~~~~~
key.c:208:13: error: 'DSA_do_sign' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  208 |             DSA_SIG* dsa_sig = DSA_do_sign(digest, dlen, key->dsa);
      |             ^~~~~~~
/usr/include/openssl/dsa.h:113:32: note: declared here
  113 | OSSL_DEPRECATEDIN_3_0 DSA_SIG *DSA_do_sign(const unsigned char *dgst, int dlen,
      |                                ^~~~~~~~~~~
cc1: all warnings being treated as errors
make[3]: *** [Makefile:590: libguac_common_ssh_la-key.lo] Error 1
make[3]: Leaving directory '/usr/src/guacamole-server-1.4.0/src/common-ssh'
make[2]: *** [Makefile:616: all-recursive] Error 1
make[2]: Leaving directory '/usr/src/guacamole-server-1.4.0/src/common-ssh'
make[1]: *** [Makefile:536: all-recursive] Error 1
make[1]: Leaving directory '/usr/src/guacamole-server-1.4.0'
make: *** [Makefile:458: all] Error 2
root@gcm-template:/usr/src/guacamole-server-1.4.0# `

Happy to persevere with 22.10 if there is benefit in doing so.

Thanks

David

**Comment by dhardyuk** [comment]
I've rerun the setup with Ubuntu 22.04 and get the same errors.

The Ubuntu version in the howto is not available in Proxmox 7.2-11 which I'm running (non ubuntu releases have been clipped from this list):

root@pve1:~# pveam available --section system
system          ubuntu-18.04-standard_18.04.1-1_amd64.tar.gz
system          ubuntu-20.04-standard_20.04-1_amd64.tar.gz
system          ubuntu-22.04-standard_22.04-1_amd64.tar.zst
system          ubuntu-22.10-standard_22.10-1_amd64.tar.zst

I'll go again without using apt upgrade and see where that gets me ......

**Comment by dhardyuk** [comment]
OK, same errors with 22.04 without apt upgrade :-(

I'll go again with 20.04!

**Comment by dhardyuk** [comment]
Yup, that was better - I got further than with the other templates but still having issues with OpenSSL as before

**Comment by dhardyuk** [comment]
I gave up.

I found a different guide to setting up Guacamole on a Debian 10 LXC container on Proxmox by helperkeys on Github which didn't defy me as much as Ubuntu did.

**Comment by obuno** [comment]
Hi there, sorry for the late reply / I never faced your issue I'm afraid.
